Mathlib/NumberTheory/Cyclotomic/Three.lean

```diff
@@ -109,3 +113,79 @@ theorem eq_one_or_neg_one_of_unit_of_congruent (hcong : ∃ n : ℤ, λ ^ 2 ∣
     have : (hζ.pow_of_coprime 2 (by decide)).toInteger = hζ.toInteger ^ 2 := by ext; simp
     simp only [this, PNat.val_ofNat, Nat.cast_ofNat, mul_neg, Int.cast_neg, ← neg_add, ←
       sub_eq_iff_eq_add.1 hx, Units.val_neg, val_pow_eq_pow_val, IsUnit.unit_spec, neg_neg]
+
+variable (x : 𝓞 K)
+
+/-- Let `(x : 𝓞 K)`. Then we have that `λ` divides one amongst `x`, `x - 1` and `x + 1`. -/
+lemma lambda_dvd_or_dvd_sub_one_or_dvd_add_one : λ ∣ x ∨ λ ∣ x - 1 ∨ λ ∣ x + 1 := by
+  classical
+  have := hζ.finite_quotient_toInteger_sub_one (by decide)
+  let _ := Fintype.ofFinite (𝓞 K ⧸ Ideal.span {λ})
+  have := Finset.mem_univ (Ideal.Quotient.mk (Ideal.span {λ}) x)
+  rw [Finset.univ_of_card_eq_three] at this
+  · simp only [Finset.mem_insert, Finset.mem_singleton] at this
+    rcases this with (h | h | h)
+    · left
+      exact Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 h
+    · right; left
+      refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
+      rw [RingHom.map_sub, h, RingHom.map_one, sub_self]
+    · right; right
+      refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
+      rw [RingHom.map_add, h, RingHom.map_one, add_left_neg]
+  · rw [← Nat.card_eq_fintype_card, hζ.card_quotient_toInteger_sub_one (by decide),
+      hζ.norm_toInteger_sub_one_of_prime_ne_two' (by decide)]
+    simp
```

Comment:
```suggestion
  let _ : Ring (𝓞 K ⧸ Ideal.span {λ}) := CommRing.toRing -- to speed up instance synthesis
  let _ : AddGroup (𝓞 K ⧸ Ideal.span {λ}) := AddGroupWithOne.toAddGroup -- dito
  have := Finset.mem_univ (Ideal.Quotient.mk (Ideal.span {λ}) x)
  have h3 : Fintype.card (𝓞 K ⧸ Ideal.span {λ}) = 3 := by
    rw [← Nat.card_eq_fintype_card, hζ.card_quotient_toInteger_sub_one (by decide),
      hζ.norm_toInteger_sub_one_of_prime_ne_two' (by decide)]
    simp only [PNat.val_ofNat, Nat.cast_ofNat, Int.reduceAbs]
  rw [Finset.univ_of_card_eq_three h3] at this
  simp only [Finset.mem_insert, Finset.mem_singleton] at this
  rcases this with (h | h | h)
  · left
    exact Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 h
  · right; left
    refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
    rw [RingHom.map_sub, h, RingHom.map_one, sub_self]
  · right; right
    refine Ideal.mem_span_singleton.1 <| Ideal.Quotient.eq_zero_iff_mem.1 ?_
    rw [RingHom.map_add, h, RingHom.map_one, add_left_neg]
```
This one was quite slow. Adding the two explicit instances makes it reasonably fast.